cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/IRConstruction.qll

```diff
@@ -166,33 +157,42 @@ private module Cached {
     // last instruction(s) in the body. A for loop may not have a condition, in
     // which case `getFirstConditionInstruction` returns the body instead.
     exists(TranslatedForStmt s |
-      s instanceof TranslatedForStmt and
-      result = s.getFirstConditionInstruction() and
-      exists(TranslatedElement inLoop, InstructionTag tag |
-        result = inLoop.getInstructionSuccessor(tag, kind) and
-        exists(TranslatedElement bodyOrUpdate |
-          bodyOrUpdate = s.getBody()
-          or
-          bodyOrUpdate = s.getUpdate()
-        |
-          inLoop = bodyOrUpdate.getAChild*()
-        ) and
-        instruction = inLoop.getInstruction(tag)
+      targetInstruction = s.getFirstConditionInstruction() and
+      targetInstruction = sourceElement.getInstructionSuccessor(sourceTag, kind) and
+      (
+        requiredAncestor = s.getUpdate()
+        or
+        not exists(s.getUpdate()) and
```

Comment:
Well spotted. I should have written a remark about this.

The original code was a bit lazy because it included edges from the body to the loop condition in all cases, even when there was an update present. I believe there can't possibly be a CFG edge from the body to the loop condition when there's an update present, and I've checked on at least one large snapshot that this is the case in practice.

I made this change to avoid the confusion and inefficiency that could result from having one source/target pair be associated with more than one `requiredAncestor` in `backEdgeCandidate`.

Reply:
Sounds good to me.